File: sdk/typescript/src/providers/anthropic.ts
/**
 * Anthropic provider monkey-patch.
 *
 * Target: `Anthropic.messages.create`
 * Min version: @anthropic-ai/sdk >= 0.18
 * Response token path: `response.usage.input_tokens`, `response.usage.output_tokens`
 * Model path: `response.model` (or request kwarg `model`)
 */

import { trace } from "@opentelemetry/api";
import * as attributes from "../attributes.js";
import { getPipelineId, getStage } from "../context.js";
import type { PricingModel } from "../pricing.js";

const MIN_VERSION = "0.18";
const PROVIDER_NAME = "anthropic";

const tracer = trace.getTracer("hikari");

interface AnthropicUsage {
  input_tokens?: number;
  output_tokens?: number;
}

interface AnthropicResponse {
  usage?: AnthropicUsage;
  model?: string;
}

const originals = new Map<string, unknown>();

function extractTokens(response: AnthropicResponse): [number | null, number | null] {
  const usage = response.usage;
  if (!usage) {
    return [null, null];
  }
  return [usage.input_tokens ?? null, usage.output_tokens ?? null];
}

function extractModel(response: AnthropicResponse, kwargs: Record<string, unknown>): string {
  return response.model ?? (kwargs.model as string) ?? "unknown";
}

function makeWrapper(
  original: (...args: unknown[]) => Promise<AnthropicResponse>,
  pricing: PricingModel
): (...args: unknown[]) => Promise<AnthropicResponse> {
  return async function (this: unknown, ...args: unknown[]): Promise<AnthropicResponse> {
    const spanName = "anthropic.messages.create";
    const stage = getStage() ?? spanName;

    return tracer.startActiveSpan(spanName, async (span) => {
      try {
        const response = await original.apply(this, args);

        try {
          const [inputTokens, outputTokens] = extractTokens(response);
          const kwargs = (args[0] as Record<string, unknown>) ?? {};
          const model = extractModel(response, kwargs);
          const pipelineId = getPipelineId();

          if (pipelineId) {
            span.setAttribute(attributes.PIPELINE_ID, pipelineId);
          }
          span.setAttribute(attributes.STAGE, stage);
          span.setAttribute(attributes.MODEL, model);
          span.setAttribute(attributes.PROVIDER, PROVIDER_NAME);

          if (inputTokens !== null) {
            span.setAttribute(attributes.TOKENS_INPUT, inputTokens);
          }
          if (outputTokens !== null) {
            span.setAttribute(attributes.TOKENS_OUTPUT, outputTokens);
          }

          const { inputCost, outputCost, totalCost } = pricing.computeCost(
            PROVIDER_NAME,
            model,
            inputTokens,
            outputTokens
          );

          if (inputCost !== null) {
            span.setAttribute(attributes.COST_INPUT, inputCost);
          }
          if (outputCost !== null) {
            span.setAttribute(attributes.COST_OUTPUT, outputCost);
          }
          if (totalCost !== null) {
            span.setAttribute(attributes.COST_TOTAL, totalCost);
          }
        } catch (err) {
          console.debug("Failed to set span attributes:", err);
        }

        span.end();
        return response;
      } catch (err) {
        span.end();
        throw err;
      }
    });
  };
}

/**
 * Apply monkey-patch to Anthropic client methods.
 *
 * @param pricing - The pricing model to use for cost computation
 * @returns `true` if patch was applied, `false` if skipped
 */
export function patch(pricing: PricingModel): boolean {
  try {
    // Dynamic import to avoid hard dependency
    const anthropic = require("@anthropic-ai/sdk");
    const version = anthropic.VERSION ?? anthropic.default?.VERSION ?? "0.0.0";

    // Version check
    const [major, minor] = version.split(".").map((v: string) => parseInt(v, 10));
    const [minMajor, minMinor] = MIN_VERSION.split(".").map((v) => parseInt(v, 10));

    if ((major ?? 0) < (minMajor ?? 0) || ((major ?? 0) === (minMajor ?? 0) && (minor ?? 0) < (minMinor ?? 0))) {
      console.warn(`@anthropic-ai/sdk version ${version} < ${MIN_VERSION}, skipping`);
      return false;
    }

    // Patch the Anthropic class prototype
    const Anthropic = anthropic.default ?? anthropic;
    const proto = Anthropic.prototype;

    if (!proto?.messages) {
      console.warn("Anthropic client structure not recognized, skipping");
      return false;
    }

    const original = proto.messages.create.bind(proto.messages);
    originals.set("create", original);

    proto.messages.create = makeWrapper(
      original as (...args: unknown[]) => Promise<AnthropicResponse>,
      pricing
    );

    console.info(`Patched @anthropic-ai/sdk ${version}`);
    return true;
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === "MODULE_NOT_FOUND") {
      console.debug("@anthropic-ai/sdk not installed, skipping");
      return false;
    }
    console.warn("Failed to patch @anthropic-ai/sdk:", err);
    return false;
  }
}

/**
 * Restore original Anthropic methods.
 */
export function unpatch(): void {
  try {
    const anthropic = require("@anthropic-ai/sdk");
    const Anthropic = anthropic.default ?? anthropic;
    const proto = Anthropic.prototype;

    const original = originals.get("create");
    if (original && proto?.messages) {
      proto.messages.create = original;
      originals.delete("create");
    }
  } catch {
    // Ignore if anthropic not installed
  }
}

File: sdk/python/src/hikari/providers/anthropic.py
"""Anthropic provider monkey-patch.

Target: ``anthropic.resources.messages.Messages.create``
         ``anthropic.resources.messages.AsyncMessages.create``
Min version: anthropic >= 0.18
Response token path: ``response.usage.input_tokens``, ``response.usage.output_tokens``
Model path: ``response.model`` (or request kwarg ``model``)
"""

from __future__ import annotations

import functools
import logging
from typing import Any, Callable

from opentelemetry import trace

from hikari import attributes
from hikari.context import get_pipeline_id, get_stage

logger = logging.getLogger("hikari.providers.anthropic")

MIN_VERSION = "0.18"
PROVIDER_NAME = "anthropic"
_originals: dict[str, Any] = {}

tracer = trace.get_tracer("hikari")


def _extract_tokens(response: Any) -> tuple[int | None, int | None]:
    usage = getattr(response, "usage", None)
    if usage is None:
        return (None, None)
    return (
        getattr(usage, "input_tokens", None),
        getattr(usage, "output_tokens", None),
    )


def _extract_model(response: Any, kwargs: dict[str, Any]) -> str:
    model = getattr(response, "model", None)
    if model is None:
        model = kwargs.get("model", "unknown")
    return model


def _make_sync_wrapper(
    original: Callable[..., Any],
    pricing_model: Any,
) -> Callable[..., Any]:
    @functools.wraps(original)
    def wrapper(self: Any, *args: Any, **kwargs: Any) -> Any:
        span_name = "anthropic.messages.create"
        stage = get_stage() or span_name
        with tracer.start_as_current_span(span_name) as span:
            try:
                response = original(self, *args, **kwargs)
            except Exception:
                raise
            else:
                try:
                    input_tokens, output_tokens = _extract_tokens(response)
                    model = _extract_model(response, kwargs)
                    pipeline_id = get_pipeline_id()
                    if pipeline_id:
                        span.set_attribute(attributes.PIPELINE_ID, pipeline_id)
                    span.set_attribute(attributes.STAGE, stage)
                    span.set_attribute(attributes.MODEL, model)
                    span.set_attribute(attributes.PROVIDER, PROVIDER_NAME)
                    if input_tokens is not None:
                        span.set_attribute(attributes.TOKENS_INPUT, input_tokens)
                    if output_tokens is not None:
                        span.set_attribute(attributes.TOKENS_OUTPUT, output_tokens)

                    input_cost, output_cost, total_cost = pricing_model.compute_cost(
                        PROVIDER_NAME, model, input_tokens, output_tokens
                    )
                    if input_cost is not None:
                        span.set_attribute(attributes.COST_INPUT, input_cost)
                    if output_cost is not None:
                        span.set_attribute(attributes.COST_OUTPUT, output_cost)
                    if total_cost is not None:
                        span.set_attribute(attributes.COST_TOTAL, total_cost)
                except Exception:
                    logger.debug("Failed to set span attributes", exc_info=True)
                return response

    return wrapper


def _make_async_wrapper(
    original: Callable[..., Any],
    pricing_model: Any,
) -> Callable[..., Any]:
    @functools.wraps(original)
    async def wrapper(self: Any, *args: Any, **kwargs: Any) -> Any:
        span_name = "anthropic.messages.create"
        stage = get_stage() or span_name
        with tracer.start_as_current_span(span_name) as span:
            try:
                response = await original(self, *args, **kwargs)
            except Exception:
                raise
            else:
                try:
                    input_tokens, output_tokens = _extract_tokens(response)
                    model = _extract_model(response, kwargs)
                    pipeline_id = get_pipeline_id()
                    if pipeline_id:
                        span.set_attribute(attributes.PIPELINE_ID, pipeline_id)
                    span.set_attribute(attributes.STAGE, stage)
                    span.set_attribute(attributes.MODEL, model)
                    span.set_attribute(attributes.PROVIDER, PROVIDER_NAME)
                    if input_tokens is not None:
                        span.set_attribute(attributes.TOKENS_INPUT, input_tokens)
                    if output_tokens is not None:
                        span.set_attribute(attributes.TOKENS_OUTPUT, output_tokens)

                    input_cost, output_cost, total_cost = pricing_model.compute_cost(
                        PROVIDER_NAME, model, input_tokens, output_tokens
                    )
                    if input_cost is not None:
                        span.set_attribute(attributes.COST_INPUT, input_cost)
                    if output_cost is not None:
                        span.set_attribute(attributes.COST_OUTPUT, output_cost)
                    if total_cost is not None:
                        span.set_attribute(attributes.COST_TOTAL, total_cost)
                except Exception:
                    logger.debug("Failed to set span attributes", exc_info=True)
                return response

    return wrapper


def patch(pricing_model: Any) -> bool:
    """Apply monkey-patch to Anthropic client methods."""
    try:
        import anthropic  # noqa: F811
    except ImportError:
        logger.debug("anthropic not installed, skipping")
        return False

    version_str = getattr(anthropic, "__version__", "0.0.0")
    try:
        from packaging.version import Version

        if Version(version_str) < Version(MIN_VERSION):
            logger.warning("anthropic version %s < %s, skipping", version_str, MIN_VERSION)
            return False
    except ImportError:
        parts = version_str.split(".")
        major, minor = int(parts[0]), int(parts[1]) if len(parts) > 1 else 0
        min_parts = MIN_VERSION.split(".")
        min_major, min_minor = int(min_parts[0]), int(min_parts[1]) if len(min_parts) > 1 else 0
        if (major, minor) < (min_major, min_minor):
            logger.warning("anthropic version %s < %s, skipping", version_str, MIN_VERSION)
            return False

    try:
        from anthropic.resources.messages import AsyncMessages, Messages

        _originals["sync_create"] = Messages.create
        _originals["async_create"] = AsyncMessages.create

        Messages.create = _make_sync_wrapper(Messages.create, pricing_model)  # type: ignore[assignment]
        AsyncMessages.create = _make_async_wrapper(AsyncMessages.create, pricing_model)  # type: ignore[assignment]
        logger.info("Patched anthropic %s", version_str)
        return True
    except Exception:
        logger.warning("Failed to patch anthropic", exc_info=True)
        return False


def unpatch() -> None:
    """Restore original Anthropic methods."""
    try:
        from anthropic.resources.messages import AsyncMessages, Messages

        if "sync_create" in _originals:
            Messages.create = _originals.pop("sync_create")  # type: ignore[assignment]
        if "async_create" in _originals:
            AsyncMessages.create = _originals.pop("async_create")  # type: ignore[assignment]
    except ImportError:
        pass
